### ISSUE TITLE
Implement operator interface in `aggregation`

### ISSUE THREAD

**Issue by AlekSi** [FerretDB/FerretDB#2317]
Things to test:
* `null`;
* missing fields;
* arrays;
* dot notation;
* numbers.

DoD:
- Create issues for aggregation operators refactoring
- Create operator interface in a separate PR, document difference
- Use this interface to show how it's used (we can return `not implemented`)
- Create a separate issue for `$type` aggregation operator

----

For the issue for `$type`:

 What should be done?

https://www.mongodb.com/docs/manual/reference/operator/aggregation/type/

The simplest approach is to do the same what we did for group's `count` and `sum`.

- Check how `$type` should be supported in find's projection and create a separate issue for it if needed
- if some parts are too complicated (e.g. dot notation), create a separate issue
- In a separate PR: move `PathExpression` that is only used by aggregation to the aggregation package. So that we don't mix it with non-aggregation part of code. If possible, rename `PathExperssion` to something more reasonable. If `type Expression interface` is not needed, remove it. Otherwise, document why it's needed. Check the need for `Null` [here](https://github.com/FerretDB/FerretDB/blob/main/internal/types/expression.go#L146-L147). If it's really needed, document why.

**Comment by w84thesun** [comment]
Blocked by #2382.

**Comment by rumyantseva** [comment]
Example for `find` (thanks @w84thesun):

```
db.values.find({}, {'v': {'$type': 'v'}})
[
  { _id: 'double-big', v: 'string' },
  { _id: 'double-2', v: 'string' },
  { _id: 'int32-min', v: 'string' },
  { _id: 'int64-double-big', v: 'string' },
]

db.values.find({v: {'$type': 'string'}}, {'fieldValue': {'$type': 'v'}})
[
  { _id: 'string-empty', fieldValue: 'string' },
  { _id: 'string-whole', fieldValue: 'string' },
  { _id: 'string', fieldValue: 'string' },
  { _id: 'string-double', fieldValue: 'string' },
  { _id: 'array-three-reverse', fieldValue: 'string' },
  { _id: 'array-three', fieldValue: 'string' },
  { _id: 'array-strings-desc', fieldValue: 'string' },
  { _id: 'array-two', fieldValue: 'string' },
  { _id: 'array-composite',
```

**Comment by rumyantseva** [comment]
@b1ron as agreed, please check how `$type` works in those two `find` cases and `$type` in aggregation.

**Comment by b1ron** [comment]
The behavior for the `$type` operator in the projection stage seems to be same with the find command. They return the same results. However, the [above](https://github.com/FerretDB/FerretDB/issues/2317#issuecomment-1547518691) syntax is wrong. To reference a field from the input document you must prefix it with the $ symbol. Otherwise it will just set the field to the type of the expression passed to the `$type` operator.

Consider the following documents. This illustrates the consistent behavior in projection, _only_.

```js
db.foo.insertMany([
    {_id: 1, a: 1},
    {_id: 2, a: '1'},
    {_id: 3, a: [1, null]}, // array
    {_id: new ObjectId()}, // missing field but has an object id
    {_id: 5, a: {b: 1}}, // sub-document
    {_id: 6, a: null}, // null
    {_id: 7, a: 1, b: 1}, // multiple fields
    {_id: 8, b: null}, // different type on field b
    {_id: 9, a: NumberDecimal(1), b: NumberLong(2), c: NumberInt(3)} // different number types
]);

const A = db.foo.find({}).projection({a: {$type: '$a'}});
const B = db.foo.aggregate([{$project: {a: {$type: '$a'}}}]);

assert.sameMembers(A.toArray(), B.toArray());
```

**Comment by w84thesun** [comment]
> To reference a field from the input document you must prefix it with the $ symbol. Otherwise it will just set the field to the type of the expression passed to the `$type` operator.

Could you please add some examples? I'm feeling that I don't get how it should work. 🤔

**Comment by w84thesun** [comment]
Oh, I see.

```js
test> db.values.find({}, {'v': {'$type': 'v'}})
[
  { _id: 'null', v: 'string' },
  { _id: 'regex', v: 'string' },
  { _id: 'double-smallest', v: 'string' },
  { _id: 'string-double', v: 'string' },
  { _id: 'int64-big', v: 'string' },
  { _id: 'regex-empty', v: 'string' },
]
```

```js
test> db.values.find({}, {'v': {'$type': '$v'}})
[
  { _id: 'null', v: 'null' },
  { _id: 'regex', v: 'regex' },
  { _id: 'double-smallest', v: 'double' },
  { _id: 'string-double', v: 'string' },
  { _id: 'int64-big', v: 'long' },
  { _id: 'regex-empty', v: 'regex' },
  { _id: 'int64-max', v: 'long' },
```
